Add dev environment

andresmgot opened the pull request:
**Issue Ref**: #426
 
**Description**: 

This PR adds a docker environment in which all the tools required to build and test Kubeless are already installed. For running the development environment `minikube` is required. To make it easier to start the dev environment the script `./script/start-test-environment.sh` sets up the environment. 

It removes the usage of the command `timeout` in the integration tests since it is only available in debian-based distributions.

**TODOs**:
 - [X] Ready to review
 ~~- [] Automated Tests~~ (Tested manually)
 - [X] Docs

phsiao commented:
Test this on a Mac, not perfect but looks better.   Have not had a chance to look into the cause of the failures yet.

```
bash-4.3# make integration-tests
./script/integration-tests
/usr/local/bin/kubectl
NAME       STATUS    ROLES     AGE       VERSION
minikube   Ready     <none>    4m        v1.8.0
Error from server (AlreadyExists): object is being deleted: namespaces "kubeless" already exists
1..34
ok 1 Verify TEST_CONTEXT envvar
ok 2 Verify needed kubernetes tools installed
ok 3 Verify minikube running (if context=='minikube')
ok 4 Verify k8s RBAC mode
ok 5 Test simple function failure without RBAC rules
ok 6 Test simple function success with proper RBAC rules
ok 7 Test function: get-python
ok 8 Test function: get-python-deps
ok 9 Test function: scheduled-get-python (part I)
ok 10 Test function update: get-python
ok 11 Test function ingress: get-python
not ok 12 Test function autoscale: get-python
# (from function `test_kubeless_autoscale' in file tests/../script/libtest.bash, line 307,
#  in test file tests/integration-tests.bats, line 57)
#   `test_kubeless_autoscale get-python' failed
# time="2017-11-28T18:51:40Z" level=fatal msg="Can't create autoscale: the server could not find the requested resource"
ok 13 Test function: get-nodejs
ok 14 Test function: get-nodejs-deps
ok 15 Test function: timeout-nodejs
ok 16 Test function: get-nodejs-multi
ok 17 Test function: get-ruby
ok 18 Test function: get-ruby-deps
ok 19 # skip (This test is flaky until kubeless/kubeless/issues/395 is fixed) Test function: get-dotnetcore
not ok 20 Test function: post-python
# (from function `verify_function' in file tests/../script/libtest.bash, line 269,
#  from function `test_kubeless_function' in file tests/../script/libtest.bash, line 274,
#  in test file tests/integration-tests.bats, line 82)
#   `test_kubeless_function post-python' failed with status 2
# time="2017-11-28T18:53:35Z" level=info msg="Deploying function..."
# time="2017-11-28T18:53:35Z" level=info msg="Function post-python submitted for deployment"
# time="2017-11-28T18:53:35Z" level=info msg="Check the deployment status executing 'kubeless function ls post-python'"
# curl --data '{"hello":"world"}' localhost:8080/api/v1/proxy/namespaces/default/services/post-python/ --header "Content-Type:application/json"
# time="2017-11-28T18:53:40Z" level=error
# time="2017-11-28T18:53:40Z" level=fatal msg="an error on the server (\"Error: 'dial tcp 172.17.0.17:8080: getsockopt: connection refused'\nTrying to reach: 'http://172.17.0.17:8080/'\") has prevented the request from succeeding"
# make[1]: *** [Makefile:137: post-python-verify] Error 1
ok 21 Test function: post-nodejs
ok 22 Test function: post-ruby
ok 23 # skip (This test is flaky until kubeless/kubeless/issues/395 is fixed) Test function: post-dotnetcore
ok 24 Test function: get-python-metadata
not ok 25 Test function: pubsub-python
# (from function `verify_function' in file tests/../script/libtest.bash, line 269,
#  from function `test_kubeless_function' in file tests/../script/libtest.bash, line 274,
#  in test file tests/integration-tests.bats, line 98)
#   `test_kubeless_function pubsub-python' failed with status 2
# Created topic "19196".
# pod "kafka-0" annotated
# Created topic "s3-python".
# time="2017-11-28T18:54:25Z" level=info msg="Deploying function..."
# time="2017-11-28T18:54:25Z" level=info msg="Function pubsub-python submitted for deployment"
# time="2017-11-28T18:54:25Z" level=info msg="Check the deployment status executing 'kubeless function ls pubsub-python'"
# make[1]: *** [Makefile:170: pubsub-python-verify] Error 1
ok 26 Test function: pubsub-python34
ok 27 Test function update: pubsub-python
ok 28 Test function: pubsub-nodejs
ok 29 Test function: pubsub-ruby
ok 30 Test function: scheduled-get-python (part II)
ok 31 Test topic list
ok 32 Test topic deletion
ok 33 Test custom runtime image
ok 34 Verify Kafka after restart (if context=='minikube')
```

andresmgot commented:
For fixing test 12 (autoscale) you need to use minikube 0.22.3 (autoscale feature doesn't work yet for kubernetes 1.8 until #399 is fixed). I assume you are using minikube 0.23.0.

I fixed 25 in the third commit of this PR so it shouldn't be a problem. I have no clue about 20 but it is probably caused because of a race condition.

I updated the PR to improve the error message and ensure that the version of minikube is the good one. It works on my Mac with minikube 0.22.3:

```
ok 1 Verify TEST_CONTEXT envvar
ok 2 Verify needed kubernetes tools installed
ok 3 Verify k8s RBAC mode
ok 4 Test simple function failure without RBAC rules
ok 5 Test simple function success with proper RBAC rules
ok 6 Test function: get-python
ok 7 Test function: get-python-deps
ok 8 Test function: scheduled-get-python (part I)
ok 9 Test function update: get-python
ok 10 Test function ingress: get-python
ok 11 Test function autoscale: get-python
ok 12 Test function: get-nodejs
ok 13 Test function: get-nodejs-deps
ok 14 Test function: timeout-nodejs
ok 15 Test function: get-nodejs-multi
ok 16 Test function: get-ruby
ok 17 Test function: get-ruby-deps
ok 18 # skip (This test is flaky until kubeless/kubeless/issues/395 is fixed) Test function: get-dotnetcore
ok 19 Test function: post-python
ok 20 Test function: post-nodejs
ok 21 Test function: post-ruby
ok 22 # skip (This test is flaky until kubeless/kubeless/issues/395 is fixed) Test function: post-dotnetcore
ok 23 Test function: get-python-metadata
ok 24 Test function: pubsub-python
ok 25 Test function: pubsub-python34
ok 26 Test function update: pubsub-python
ok 27 Test function: pubsub-nodejs
ok 28 Test function: pubsub-ruby
ok 29 Test function: scheduled-get-python (part II)
ok 30 Test topic list
ok 31 Test topic deletion
ok 32 Test custom runtime image
ok 33 Verify Kafka after restart (if context=='minikube')
```

andresmgot commented:
@phsiao did you have a change to test the latest changes?

phsiao commented:
Yep, works great for me too once I install minikube 0.22.3.

```
# make integration-tests
./script/integration-tests
/usr/local/bin/kubectl
NAME       STATUS    ROLES     AGE       VERSION
minikube   Ready     <none>    14m       v1.7.5
clusterrolebinding "kube-dns-admin" created
namespace "kubeless" created
1..33
ok 1 Verify TEST_CONTEXT envvar
ok 2 Verify needed kubernetes tools installed
ok 3 Verify k8s RBAC mode
ok 4 Test simple function failure without RBAC rules
ok 5 Test simple function success with proper RBAC rules
ok 6 Test function: get-python
ok 7 Test function: get-python-deps
ok 8 Test function: scheduled-get-python (part I)
ok 9 Test function update: get-python
ok 10 Test function ingress: get-python
ok 11 Test function autoscale: get-python
ok 12 Test function: get-nodejs
ok 13 Test function: get-nodejs-deps
ok 14 Test function: timeout-nodejs
ok 15 Test function: get-nodejs-multi
ok 16 Test function: get-ruby
ok 17 Test function: get-ruby-deps
ok 18 # skip (This test is flaky until kubeless/kubeless/issues/395 is fixed) Test function: get-dotnetcore
ok 19 Test function: post-python
ok 20 Test function: post-nodejs
ok 21 Test function: post-ruby
ok 22 # skip (This test is flaky until kubeless/kubeless/issues/395 is fixed) Test function: post-dotnetcore
ok 23 Test function: get-python-metadata
ok 24 Test function: pubsub-python
ok 25 Test function: pubsub-python34
ok 26 Test function update: pubsub-python
ok 27 Test function: pubsub-nodejs
ok 28 Test function: pubsub-ruby
ok 29 Test function: scheduled-get-python (part II)
ok 30 Test topic list
ok 31 Test topic deletion
ok 32 Test custom runtime image
ok 33 Verify Kafka after restart (if context=='minikube')
```

andresmgot commented:
Cool, thanks for checking
